src/query_builder.rs

```diff
@@ -0,0 +1,737 @@
+        let values = ::std::mem::replace(&mut self.values, Default::default());
+        let types = ::std::mem::replace(&mut self.types, Default::default());
+        let query_inputs = QueryInputs::new(types, values)?;
+        let read = self.store.begin_read()?;
+        read.q_once(&self.sql, query_inputs)
+    }
+
+    pub fn execute_scalar(mut self) -> Result<ScalarResult> {
+        let results = self.execute()?;
+        let scalar = ScalarResult(results.into_scalar()?);
+        Ok(scalar)
+    }
+
+    pub fn execute_coll(mut self) -> Result<CollResult> {
+        let results = self.execute()?;
+        let coll = CollResult(results.into_coll()?);
+        Ok(coll)
+    }
+
+    pub fn execute_tuple(mut self) -> Result<TupleResult> {
+        let results = self.execute()?;
+        let find_spec = results.spec.columns().map(|e| format!("{}", e)).collect();
+        let tuple = TupleResult::new(&find_spec, results.into_tuple()?.unwrap_or(vec![]));
+        Ok(tuple)
+    }
+
+    pub fn execute_rel(mut self) -> Result<RelResult> {
+        let results = self.execute()?;
+        let find_spec = results.spec.columns().map(|e| format!("{}", e)).collect();
+        let tuple = RelResult::new(find_spec, results.into_rel()?);
+        Ok(tuple)
+    }
+}
+
+pub trait VarResultsRow {
+    fn var_as_entid(&self, var: &str) -> Option<Entid>;
+    fn var_as_kw(&self, var: &str) -> Option<Rc<NamespacedKeyword>>;
+    fn var_as_boolean(&self, var: &str) -> Option<bool>;
+    fn var_as_long(&self, var: &str) -> Option<i64>;
+    fn var_as_double(&self, var: &str) -> Option<f64>;
+    fn var_as_instant(&self, var: &str) -> Option<DateTime<Utc>>;
+    fn var_as_timestamp(&self, var: &str) -> Option<i64>;
+    fn var_as_string(&self, var: &str) -> Option<Rc<String>>;
+    fn var_as_uuid(&self, var: &str) -> Option<Uuid>;
+    fn var_as_uuid_str(&self, var: &str) -> Option<String>;
+}
+
+pub trait ColResultsRow {
+    fn col_as_entid(&self, col: usize) -> Option<Entid>;
+    fn col_as_kw(&self, col: usize) -> Option<Rc<NamespacedKeyword>>;
+    fn col_as_boolean(&self, col: usize) -> Option<bool>;
+    fn col_as_long(&self, col: usize) -> Option<i64>;
+    fn col_as_double(&self, col: usize) -> Option<f64>;
+    fn col_as_instant(&self, col: usize) -> Option<DateTime<Utc>>;
+    fn col_as_timestamp(&self, col: usize) -> Option<i64>;
+    fn col_as_string(&self, col: usize) -> Option<Rc<String>>;
+    fn col_as_uuid(&self, col: usize) -> Option<Uuid>;
+    fn col_as_uuid_string(&self, col: usize) -> Option<String>;
+}
+
+pub struct ScalarResult(Option<TypedValue>);
+
+impl TypedValueConverter for ScalarResult {
+    fn as_entid(mut self) -> Option<Entid> {
+        self.0.take().map_or(None, |t| t.as_entid())
+    }
+
+    fn as_kw(mut self) -> Option<Rc<NamespacedKeyword>> {
+        self.0.take().map_or(None, |t| t.as_kw())
+    }
+
+    fn as_boolean(mut self) -> Option<bool> {
+        self.0.take().map_or(None, |t| t.as_boolean())
```

Comment:
By using `take` you make it impossible to do this:

```
let result = query.execute_scalar()?;
if let Some(yes) = result.as_boolean() {
  …
} else if let Some(long) = result.as_long() {
  …
}
```

Again, this comes back to `as`/`into`/`to` distinctions. These `Copy` types can all be `to_boolean` and use things like `self.0.map(|t| t.to_boolean())`. The non-copy types probably want `as` and `into` versions.